### PR TITLE
Add linting rule for nested type check

### PR DESCRIPTION
**Pull request by liuting1014**

### CODE REVIEW

File: src/rules/no-nested-type-without-ref.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        getAllObjectProperties(json, 'components.schemas').forEach(({ object }) => {
+            if (hasProperty(object, 'properties')) {
+               getAllObjectProperties(object, 'properties').forEach(({path, name, object}) => {
+                    if (hasProperty(object, 'type')) {
+                        const type = getProperty(object, 'type');
+                        if (type === 'array' && hasProperty(object, 'items')) {
+                            const prop = Object.keys(getProperty(object, 'items') as any)[0];
+                            if (prop === 'type' && getProperty(getProperty(object, 'items') as any, 'type') as any === 'object') {
+                                throw new ValidationError(`Invalid nested type ${name} found in ${path}`);
+                            }
```
Comment (johanblumenberg-symphony) [inline src/rules/no-nested-type-without-ref.ts]: Looks a bit strange with all the casts to `any`, why are those needed? Is it possible to fix the types?

File: src/rules/no-nested-type-without-ref.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    validate(json: ApiSpec) {
+        getAllObjectProperties(json, 'components.schemas').forEach(({ object }) => {
+            if (hasProperty(object, 'properties')) {
+               getAllObjectProperties(object, 'properties').forEach(({path, name, object}) => {
+                    if (hasProperty(object, 'type')) {
+                        const type = getProperty(object, 'type');
+                        if (type === 'array' && hasProperty(object, 'items')) {
+                            const prop = Object.keys(getProperty(object, 'items') as any)[0];
```
Comment (johanblumenberg-symphony) [inline src/rules/no-nested-type-without-ref.ts]: Looks like it only checks the first property?

`getProeprty(object, 'items.type')` ?

Comment (liuting1014) [reply src/rules/no-nested-type-without-ref.ts]: `getProperty(object, 'items.type')` doesn't work, but I updated it in another way and it checks not only the first property now.

File: src/rules/no-nested-type-without-ref.ts (hunk 1 above)
Comment (liuting1014) [reply src/rules/no-nested-type-without-ref.ts]: Fixed

File: src/rules/no-nested-type-without-ref.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    validate(json: ApiSpec) {
+        getAllObjectProperties(json, 'components.schemas').forEach(({ object }) => {
+            if (hasProperty(object, 'properties')) {
+               getAllObjectProperties(object, 'properties').forEach(({path, name, object}) => {
+                    if (hasProperty(object, 'type')) {
+                        const type = getProperty(object, 'type');
+                        if (type === 'array' && hasProperty(object, 'items')) {
+                            const items = getProperty(object, 'items');
+                            if (hasProperty(items as ApiItem, 'type') && getProperty(items as ApiItem, 'type') === 'object') {
```
Comment (johanblumenberg-symphony) [inline src/rules/no-nested-type-without-ref.ts]: If you check if `items` is defined, you don't need to cast to `ApiItem`

Comment (johanblumenberg-symphony) [reply src/rules/no-nested-type-without-ref.ts]: Yo falso don't need to call `hasProperty()` first, `getProperty()` will return `undefined` if the property doesn't exist, and that will not `=== `object'`

File: src/rules/no-nested-type-without-ref.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+                                throw new ValidationError(`Invalid nested type ${name} found in ${path}`);
```
Comment (johanblumenberg-symphony) [inline src/rules/no-nested-type-without-ref.ts]:
```suggestion
                                throw new ValidationError(`Nested type ${name} found in ${path}`);
```

File: src/rules/no-nested-type-without-ref.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+                            }
+                        }
+                        if (type === 'object') {
+                            throw new ValidationError(`Invalid nested type ${name} found in ${path}`);
```
Comment (johanblumenberg-symphony) [inline src/rules/no-nested-type-without-ref.ts]:
```suggestion
                            throw new ValidationError(`Nested type ${name} found in ${path}`);
```

File: src/rules/no-nested-type-without-ref.ts (hunk 4 above)
Comment (johanblumenberg-symphony) [reply src/rules/no-nested-type-without-ref.ts]: The type is not really invalid, the problem is that it is nested

**Review by johanblumenberg-symphony — APPROVED** [review]
